kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/LogCompactionWriterSuite.scala

```diff
@@ -0,0 +1,281 @@
+
+import io.delta.kernel.Table
+import io.delta.kernel.data.Row
+import io.delta.kernel.defaults.utils.TestRow
+import io.delta.kernel.engine.Engine
+import io.delta.kernel.expressions.Literal
+import io.delta.kernel.hook.PostCommitHook
+import io.delta.kernel.internal.{DeltaLogActionUtils, SnapshotImpl}
+import io.delta.kernel.internal.TableConfig.TOMBSTONE_RETENTION
+import io.delta.kernel.internal.actions._
+import io.delta.kernel.internal.compaction.LogCompactionWriter
+import io.delta.kernel.internal.fs.Path
+import io.delta.kernel.internal.hook.LogCompactionHook
+import io.delta.kernel.internal.replay.ActionsIterator
+import io.delta.kernel.internal.util.FileNames.DeltaLogFileType
+import io.delta.kernel.internal.util.ManualClock
+import io.delta.kernel.internal.util.Utils.singletonCloseableIterator
+import io.delta.kernel.types.{IntegerType, StructType}
+import io.delta.kernel.utils.FileStatus
+
+import org.apache.spark.sql.delta.{DeltaLog, DomainMetadataTableFeature}
+import org.apache.spark.sql.delta.DeltaOperations.Truncate
+import org.apache.spark.sql.delta.actions.{DomainMetadata => DeltaSparkDomainMetadata, TableFeatureProtocolUtils}
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+
+/**
+ * Test suite for io.delta.kernel.internal.compaction.LogCompactionWriter
+ */
+class LogCompactionWriterSuite extends CheckpointSuiteBase {
+  val COMPACTED_SCHEMA =
+    new StructType()
+      .add("txn", SetTransaction.FULL_SCHEMA)
+      .add("add", AddFile.FULL_SCHEMA)
+      .add("remove", RemoveFile.FULL_SCHEMA)
+      .add("metaData", Metadata.FULL_SCHEMA)
+      .add("protocol", Protocol.FULL_SCHEMA)
+      // .add("cdc", new StructType())
+      .add("domainMetadata", DomainMetadata.FULL_SCHEMA);
+
+  // check if a row is all null
+  def rowIsNull(row: Row): Boolean = {
+    val schema = row.getSchema()
+    for (ordinal <- 0 until schema.length()) {
+      if (!row.isNullAt(ordinal)) {
+        return false
+      }
+    }
+    true
+  }
+
+  // Get the expected actions from the log.  We filter down to just the actions that end up in a
+  // compacted log file, and also filter out adds that have been removed as well as duplicate
+  // metadata/protocol actions
+  def getActionsFromLog(
+      tablePath: Path,
+      engine: Engine,
+      startVersion: Long,
+      endVersion: Long): Seq[TestRow] = {
+    val files = DeltaLogActionUtils.listDeltaLogFilesAsIter(
+      engine,
+      new HashSet(Arrays.asList(DeltaLogFileType.COMMIT)),
+      tablePath,
+      startVersion,
+      Optional.of(endVersion),
+      false /* mustBeRecreatable */ )
+      .toInMemoryList()
+    Collections.reverse(files) // we want things in reverse order
+    val actions =
+      new ActionsIterator(engine, files, COMPACTED_SCHEMA, Optional.empty())
+    val removed = scala.collection.mutable.HashSet.empty[String]
+    var seenMetadata = false
+    var seenProtocol = false
+    val resBuilder = Seq.newBuilder[TestRow]
+    while (actions.hasNext()) {
```

Comment:
I feel like essentially rewriting log replay here in order to test this is kind of weird. Wondering if instead we can just do a super simple example and actually check for the adds/removes/metadatas that we know we committed and should be there? open to thoughts

i.e.
```
version 0 - metadata1
version 1 - add1, add2
version 2 - remove1, add3, dm1
version 3 - dm1_2, metadata2
// etc
 
expect metadata2, dm1_2, add3, remove1, add2
```

Reply:
I felt the same. I think just a self contained tests where the expected output is hardcoded will make it easier to understand the tests.